Add presets for Electra and checkpoint conversion script

pranavvp16 opened the pull request:
I have uploaded the weights on personal google cloud bucket. The `from_preset` method works properly in my local setup, but it throws some error in google collab  [notebook](https://colab.research.google.com/drive/1K5zTjzoJrFGJ5Ej4s3g8lWzJ2VMqhgwu?usp=sharing).

mattdangerw commented:
Please format your code with `./shell/format.sh`, also looks like we have a pretty simple merge conflict to resolve.

mattdangerw commented on keras_nlp/models/electra/electra_presets.py:
```diff
@@ -0,0 +1,141 @@
+            "lowercase": False,
+        },
+        # TODO: Upload weights on GCS.
+        "weights_url": "https://storage.googleapis.com/pranav-keras/electra-base-generator/model.weights.h5",
```
We actually just moved all our weights over to Kaggle. https://github.com/keras-team/keras-nlp/releases/tag/v0.7.0

This will make it easier to upload models long term, but let me get back to you next week on exact steps for upload. If you have a kaggle username, could you reply here with it?

pranavvp16 replied:
[kaggle](https://www.kaggle.com/pranavprajapati16) here is my kaggle username

mattdangerw commented on keras_nlp/models/electra/electra_presets.py:
```diff
@@ -0,0 +1,141 @@
+"""Electra model presets and configurations."""
+
+backbone_presets = {
+    "google/electra-base-generator": {
```
Also, maybe a silly question, but what is the use case for the generator checkpoints?

I think it's probably fine to have them, but my understanding is the discriminators are the main artifact for downstream use.

Could you use the generator to further fine tune with the gan-like setup?

pranavvp16 replied:
The generator is used for Fill mask task. As per the hugging face [inference api](https://huggingface.co/google/electra-base-generator?text=The+goal+of+life+is+%5BMASK%5D.). I don't have idea regarding using generators as GAN.

mattdangerw commented on keras_nlp/models/electra/electra_preprocessor.py:
````diff
@@ -0,0 +1,165 @@
+    Directly calling the layer on data.
+    ```python
+    preprocessor = keras_nlp.models.ElectraPreprocessor.from_preset(
+        "electra-base-generator"
````
I would show the discriminator here and in all other example docstrings, if that's the most useful most for actually downstream use.

mattdangerw reviewed (changes requested):
Looks overall good! Left a few comments. See Kaggle comment below.

pranavvp16 commented:
I have updated the script conversion file according to the new kaggle format. Will I be able to upload the weights from my kaggle account as mentioned in #1422

mattdangerw commented:
@pranavvp16 yes, you should be able to create a new model (just not make it public). Please share it with matthewdwatson (me) on kaggle. Thank you!

abuelnasr0 commented:
@pranavvp16 Try to upgrade kagglehub to 0.1.9 (latest version). I was using kagglehub 0.1.5 and it wasn't working right, but I was trying to upload a model not using a model. but I was getting a similar error though.
https://github.com/Kaggle/kagglehub

pranavvp16 commented:
@abuelnasr0 I tried running the script in kaggle notebook and still failed [notebook](https://www.kaggle.com/code/pranavprajapati16/keras-electra). Was your model public while trying to load from kaggle ?

abuelnasr0 commented:
@pranavvp16 yes it worked with me. And I tried it now with a private model and it worked, but I didn't upgrade kagglehub. It worked with kagglehub=0.1.6
https://www.kaggle.com/mohamedabuelnasr/bloom-preset

pranavvp16 commented:
[notebook](https://www.kaggle.com/code/pranavprajapati16/electra-preset) I also added the model using the UI of kaggle, also I validated my credentials using `kagglehub.login()` . @mattdangerw I have added you the collaborators please can you help

pranavvp16 commented:
@mattdangerw I have linked notebooks above and my url is also the correct format for the kaggle model. [notebook](https://www.kaggle.com/code/pranavprajapati16/electra-preset).

abuelnasr0 commented:
@pranavvp16 
Try this in a kaggle notebook
```
%env  KAGGLE_USERNAME=<YOUR_USERNAME>
%env  KAGGLE_KEY=<YOUR_KEY>
```

mattdangerw commented:
Thanks @pranavvp16 could you still share the link to the model, and share it with me on kaggle? I can't find you model update to look at all the files. Here's me on Kaggle. https://www.kaggle.com/matthewdwatson

Thanks!

mattdangerw commented:
Some other notes. I don't believe that you will need to authenticate with Kaggle as long as you are logged in for your Kaggle notebook with the same user you create the Kaggle model with. For things like local testing and colab, you will need to set the `KAGGLE_USERNAME` and `KAGGLE_KEY` environment variables.

All user models are private, so it is expected that you can't access https://www.kaggle.com/models/mohamedabuelnasr/bloom, I don't believe that has been shared with you. You should be able to access your own model after proper authentication.

Soon, we will all be able to create public models, and life will get easier. The whole Keras -> Kaggle flow is still under development with some rough edges.

abuelnasr0 commented:
>Some other notes. I don't believe that you will need to authenticate with Kaggle as long as you are logged in for your Kaggle notebook with the same user you create the Kaggle model with. For things like local testing and colab, you will need to set the KAGGLE_USERNAME and KAGGLE_KEY environment variables.

@mattdangerw That should be the expected behaviour, but sometimes kagglehub library returns weird errors. 
for example I wanted the upload a model
```
handle = 'mohamedabuelnasr/bloom/keras/bloom_560m_multi'
local_model_dir = '/kaggle/working/bloom_560m_multi'
kagglehub.model_upload(handle, local_model_dir)
```
and the library returns that error
```
Model 'bloom' does not exist or access is forbidden for user 'mohamedabuelnasr'. Creating or handling Model...
```
And `Unauthorized` at the end of the error message.
Although I am logged in. but when I add the env variables, everything work right.

So I suggested to add env variables as a desperate attempt to work around any potential unexpected behaviour from the kagglehub library.

@pranavvp16 one thing that's worthy to mention also is that the models takes time to be available may be 10-15 minutes at most, so try to wait a while after uploading it, then download it.

pranavvp16 commented:
@mattdangerw I tried setting up the env variables still getting the same error. Also I have added you in collaborators with administrator permissions so you can have a look at the model

mattdangerw commented:
@pranavvp16 thanks for sharing! I can see the model now, and this looks like two path issue.

1. Your entire model path, for all variants, is `electra_base_discriminator_en`, though it looks like you have tried to rename it to `electra`. I am not sure if a rename is possible. You might  have to create a new model called `electra`. The model name != the variant name. Today, if you tried `ElectraTokenizer.from_preset("kaggle://pranavprajapati16/electra_base_discriminator_en/keras/electra_base_discriminator_en/2")` you would see your assets attach. But the proper fix is making a new model named Electra, which you should see at the link https://www.kaggle.com/models/pranavprajapati16/electra when things are working.
2.  You have an extra top-level directory which should not be there. Instead of seeing all your files inside `electra_base_generator_en`, the `config.json`, `tokenizer.json`, etc, should be at the toplevel of your directory.

I've shared a quick example for one variant... Can you see if this works for you?
https://www.kaggle.com/models/matthewdwatson/electra
https://www.kaggle.com/code/matthewdwatson/electra-preset

pranavvp16 commented:
Thanks it worked @mattdangerw. Now I will create a new model and then upload its variant and make updates to the `electra_presets.py` file. Thanks once again!

mattdangerw commented:
@pranavvp16 sorry may have missed this, but is this ready for review?

mattdangerw commented on keras_nlp/models/electra/electra_presets.py:
```diff
@@ -0,0 +1,69 @@
+                "ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators"
+                "This is base discriminator model with 12 layers."
+            ),
+            "params": "109482240",
```
This should be just an int.

mattdangerw commented on keras_nlp/models/electra/electra_presets.py:
```diff
@@ -0,0 +1,69 @@
+            "params": "109482240",
+            "official_name": "ELECTRA",
+            "path": "electra",
+            "model_card": "https://huggingface.co/google/electra-base-discriminator",
```
maybe just link to the github google repo instead. that's the closest to a original model card we have https://github.com/google-research/electra

mattdangerw commented on keras_nlp/models/electra/electra_tokenizer_test.py:
```diff
@@ -40,3 +42,21 @@ def test_lowercase(self):
+    def test_smallest_preset(self):
+        self.run_preset_test(
+            cls=ElectraTokenizer,
+            preset="distil_bert_base_en_uncased",
```
Update this to electra.

mattdangerw commented on keras_nlp/models/electra/electra_backbone_test.py:
```diff
@@ -54,3 +54,34 @@ def test_saved_model(self):
+    def test_smallest_preset(self):
+        self.run_preset_test(
+            cls=ElectraBackbone,
+            preset="electra-small-generator",
```
Update this to an actual preset name.

mattdangerw commented on keras_nlp/models/electra/electra_backbone_test.py:
```diff
@@ -54,3 +54,34 @@ def test_saved_model(self):
+                "pooled_output": (1, 256),
+            },
+            # The forward pass from a preset should be stable!
+            # TODO: Add sequence and pooled output trimmed to 5 tokens.
```
we should do this!

mattdangerw commented on keras_nlp/models/electra/electra_preprocessor.py:
```diff
@@ -0,0 +1,163 @@
+     2. Pack the inputs together using a `keras_nlp.layers.MultiSegmentPacker`.
+       with the appropriate `"[CLS]"`, `"[SEP]"` and `"[PAD]"` tokens.
+     3. Construct a dictionary of with keys `"token_ids"` and `"padding_mask"`,
+       that can be passed directly to a DistilBERT model.
```
Take a pass over docstrings for things like this that persisted from a copy paste.

mattdangerw reviewed (changes requested):
Looks good overall! Few changes.

pranavvp16 commented:
I have made all the changes as suggested

mattdangerw commented on keras_nlp/models/electra/electra_presets.py:
```diff
@@ -0,0 +1,69 @@
+            "path": "electra",
+            "model_card": "https://github.com/google-research/electra",
+        },
+        "kaggle_handle": "kaggle://pranavprajapati16/electra/keras/electra_base_discriminator_en/1",
```
I don't see anything at https://www.kaggle.com/models/pranavprajapati16/electra.

You should now have the ability to make models public, can you do so? Or is the actual model here? https://www.kaggle.com/models/pranavprajapati16/electra_base_discriminator_en (in which case these links are still wrong).

Let me know where to get the proper assets and I will copy to the Keras org.

pranavvp16 replied:
Sorry the model was private just made it public. https://www.kaggle.com/models/pranavprajapati16/electra

mattdangerw replied:
Thanks! Uploading now! I can just patch the new links into this PR and land. I'll ping here if I run into any issues.

mattdangerw reviewed (approved):
This looks great! Just let me know where the final assets to copy over are and I will pull this in.

mattdangerw commented:
Actually does look like there is an error here. It looks like the tokenizer should be configured to lowercase input, but is not. This is leading to some test failures.

E.g. `input_data=["The quick brown fox."],` -> an UNK token at the beginning of the sequence and failing tests.

Can you take a look and confirm that we should be lowercasing input for all electra presets? If so I can go ahead an make the changes here, there will be some annoying renames to stick to our conversions--we should call the variants `uncased_en` instead of `_en`.

mattdangerw commented:
Also, could you try converting the `large` versions? And adding presets? We should probably include those.

pranavvp16 commented:
regarding the tokenizer I found this [config](https://huggingface.co/google/electra-base-generator/blob/main/tokenizer_config.json) for one of the presets. Also I have uploaded the weights for large models

mattdangerw commented:
Thanks! I'll get large uploaded, and fix up the lowercasing issues.

mattdangerw commented:
OK think this is working! Going to pull this in. @pranavvp16 please let us know if you spot any issues. We should probably test this out with an end to end colab to make sure things are working as intended.

mattdangerw commented:
Not that I changed all the preset names to include "uncased" in keeping with bert conventions. `electra_small_generator_uncased_en`
